File: Haskell/Day01.hs
import AdventHelper

import Data.List
import Data.List.Split
import Data.Tuple.Extra

parseInput :: String -> (Int, Int)
parseInput s =  both read (head ss, last ss)
    where ss = splitOn " " s

part1 :: ([Int], [Int]) -> Int
part1 (lhs, rhs) = sum $ zipWith (\x y -> abs (x - y)) lhs rhs

countIf :: Eq a => [a] -> a -> Int
countIf xs x = length $ filter (== x) xs

part2 :: ([Int], [Int]) -> Int
part2 (lhs, rhs) = sum $ zipWith (*) lhs $ map (countIf rhs) lhs

main :: IO ()
main = do
    f <- readFile "../input/input01.txt"
    let s = lines f
    let q = map parseInput s
    let r = (sort $ map fst q, sort $ map snd q)
    printSoln 1 (part1 r) (part2 r)

File: C++/Day01.cpp
#ifdef DAY01

#include "AH.h"

namespace Day01
{

	void parseInput(
		const std::vector<std::string> & is,
		std::vector<int> & lhs,
		std::vector<int> & rhs
	)
	{
		lhs.reserve(is.size());
		rhs.reserve(is.size());
		for (auto & i : is) {
			const auto parts = AH::SplitOnString(i, "   ");
			lhs.push_back(std::stoi(parts[0]));
			rhs.push_back(std::stoi(parts[1]));
		}

		return;
	}

	int freq(
		const std::vector<int> & lhs,
		const std::vector<int> & rhs)
	{
		int s = 0;
		std::map<int, int> f;
		for (auto v : lhs)
		{
			if (f.count(v) == 0)
			{
				for (auto u : rhs) {
					if (u == v)
						f[v]++;
				}
			}
			s += v * f[v];
		}
		return s;
	}

	int Run(const std::string& filename)
	{
		const auto inputLines = AH::ReadTextFile(filename);
		std::vector<int> lhs, rhs;
		parseInput(inputLines, lhs, rhs);
		std::sort(lhs.begin(), lhs.end());
		std::sort(rhs.begin(), rhs.end());

		int part1 = 0;
		int part2 = freq(lhs, rhs);
		for (size_t i = 0; i <  inputLines.size(); i++)
		{
			part1 += std::abs(lhs[i] - rhs[i]);
		}

		AH::PrintSoln(1, part1, part2);

		return 0;
	}

}

#endif
